Creative commons licensing support

esbeka opened the pull request:
### Background
This pull request allows course content to be published under the Creative Commons license, with support for CC options like: BY, NC, ND, and SA. When a course is set to a certain license, the footer of each course page will hold the license information. Additionally, videos can be licensed separately as well. See [creativecommons.org/about](http://creativecommons.org/about) for more details on the CC license.

### Studio updates
Adds a feature flag to enable/disable Creative Commons licensing in general. Through the advanced course settings, you are able to disable the licensing option for a certain course.
A license selector is added to the course settings page and video edit modal. It allows the publisher to license their work under the latest Creative Commons license.
In the course outline, and underneath videos that have a license set, the license will be previewed as it would be displayed in the LMS.

### LMS updates
Adds a feature flag to enable/disable Creative Commons licensing in general. When enabled, it depicts the license underneath each course page and video element.

### Testing
When the feature flag CREATIVE_COMMONS_LICENSING is set to False in CMS and LMS, the views of the course settings editor, video edit modal, course outline page, and view pages should not be changed.
In case the feature flag is set to True, the following screens in Studio should have changed:

#### in Studio
##### Enabling CC Licensing
To enable the CC Licensing feature, make sure the CREATIVE_COMMONS_LICENSING and DEFAULT_COURSE_LICENSEABLE feature flags are set to True. In case the latter is set to False, you can also configure it per course, by navigating to the Advanced Course Settings and setting the Licensable parameter to True.
![course_advanced_settings](https://cloud.githubusercontent.com/assets/2682577/5412086/68ecf17e-8209-11e4-9e9a-b1694965ae3c.png)

##### Setting the course license to: CC
CC allows you to set the license with certain options. Attribution is set by default. Non Derivatives and Share Alike are XOR (they cannot be set at the same time).
![course_edit_cc](https://cloud.githubusercontent.com/assets/2682577/5412097/80b88aa2-8209-11e4-94ac-2f7da79245c4.png)

##### Setting the course license to: All rights reserved
![course_edit_arr](https://cloud.githubusercontent.com/assets/2682577/5412100/89d0f138-8209-11e4-95e9-322598196916.png)

##### Showing the course content license in the outline
![course_outline](https://cloud.githubusercontent.com/assets/2682577/5412106/9c8ef5c2-8209-11e4-8df7-185317204a42.png)

##### Editing the course content license in the outline
![course_outline_edit](https://cloud.githubusercontent.com/assets/2682577/5412114/ae07c2b6-8209-11e4-9093-cbfe431b455f.png)

##### Editing video license
CC allows you to set the license with certain options. Attribution is set by default. Non Derivatives and Share Alike are XOR (they cannot be set at the same time).
![video_edit](https://cloud.githubusercontent.com/assets/2682577/5412129/c3b047c8-8209-11e4-90f5-065cfda406d5.png)

##### Video license preview
![cms_video_preview](https://cloud.githubusercontent.com/assets/2682577/5412131/cf9a92e6-8209-11e4-8705-131061114a8b.png)

##### Video license preview hover
Clicking on the license will open the correct page on creativecommons.org that shows all the details on that license. Except for the All rights reserved license, this license is not clickable.
![cms_video_preview_hover](https://cloud.githubusercontent.com/assets/2682577/5412137/d8f96ede-8209-11e4-88dd-ef6f713474ba.png)

#### in LMS
The test instructions are not available yet, as this is still work in progress. I will upload those as soon as the LMS part is done. The idea is that the license will look exactly like the previewed license in the video preview. Where the course content license will be shown in the grey background area that surrounds the content.

### Discussion
**As discussed with**: @sarina, @explorerleslie, @marcotuts

**Discussion links**:
* Previous PR: #5057 
* [https://openedx.atlassian.net/browse/UX-948](JIRA UX-948)
* [https://openedx.atlassian.net/browse/OSPR-42](JIRA OSPR-42)
* [https://openedx.atlassian.net/browse/OSPR-89](JIRA OSPR-89)

esbeka commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,82 @@
+                        licenseLink.push("nd");
+                        licenseText.push(gettext("NonDerivatives"));
+                    }
+                    licenseTooltip = interpolate(gettext("This work is licensed under a Creative Commons %(license_attributes)s %(version)s International License."), {
```
Unfortunately this is not possible, I picked these names as the Creative Commons guidelines specify them. You can find an example here: http://creativecommons.org/licenses/by-nc-nd/4.0/
In this case the official license name is:
Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International

In case these get translated, the independent attributes will need to be translated as (Dutch example):
Creative Commons Naamsvermelding-NietCommercieel-GeenAfgeleideWerken 4.0 Internationaal
See: http://creativecommons.org/licenses/by-nc-nd/4.0/deed.nl

clrux replied:
Fair enough, thanks for researching @esbeka.

esbeka commented on cms/templates/course_outline.html:
```diff
@@ -100,6 +100,23 @@ <h2 class="status-release-label">${_("Course Start Date:")}</h2>
                       </li>
                     </ul>
                 </div>
+                % if settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False) and license:
```
Its the course content license, its passed to the view just like course_release_date. Should this be changed to course_content_license?

sarina commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -372,19 +373,25 @@ def format_course_for_view(course):
         Return a dict of the data which the view requires for each unsucceeded course
         """
-        return {
+        fields = {
```
was there a problem with leaving this as "return"?

esbeka replied:
Ah, good one. I missed that change, will revert it. The license and licensable attributes were optional before.

clrux commented on cms/static/sass/elements/_forms.scss:
```diff
@@ -138,6 +138,100 @@ form {
+      line-height: 1.5;
+      border-bottom: 1px solid #EEE;
+      padding: ($baseline / 2) 0 ($baseline * 0.4);
+      &.clickable {
```
We use stateful classnames. So, `clickable` (if something is clickable), should be `is-clickable`. Related, if something _has_ a state we use `has-_state_`.

marcotuts commented:
@esbeka for the shift of the License meta information from the Course Outline header to the bottom right of the page, I was looking around in the cms code to try and find where that might be added. I've written down some notes here but I'm going to tag @andy-armstrong in case he has a chance to skim what I've read here and see if he has other suggestions.

Option 1: In the browser I played around with locating this in the main content area on the outline page and you can see it here - 
![image](https://cloud.githubusercontent.com/assets/2023680/5646886/e06d8b60-964e-11e4-9650-0b59d5b3c7b6.png) I didn't drop in the full line with the icons but I dropped in an example right aligned section in the primary content column area. 

It looked to me like you'd want to add in logic to the template that would show the license possibly in a file called course-outline.underscore, adding a case for the course-meta section. That being said, this section is all wired up to be drag/droppable so perhaps if this isn't ideal there is also option 2 below.

Option 2: Add the course-meta text outside of the section class="content" and drop it into the course_outline.html template.

esbeka commented:
@marcotuts Thank you for researching these options. I've implemented option 2, it's in the latest commits.

esbeka commented:
@sarina All issues with the quality tests are resolved with the last commit. Additionally, fixed the following issues:
- [x] The course outline shows the license at the bottom as discussed with @marcotuts.
- [x] The course content license will default to All rights reserved if none is set yet (will need to test this in a new devstack, will do that later).
- [x] Fixed an issue were multiple consecutive edits of the course would only work the first time its edited (until the page was refreshed).
- [x] Made the code more readable by referring to video_license or context_course.license instead of the license variable.

sarina commented on common/static/css/edx-cc.css:
```diff
@@ -0,0 +1,63 @@
+  font-style: normal;
+}
+/* Chrome hack: SVG is rendered more smooth in Windozze. 100% magic, uncomment if you need it. */
+/* Note, that will break hinting! In other OS-es font will be not as sharp as it could be */
```
The sandbox keeps failing to build with the following 2 errors - I don't know how to resolve it. I've reached out to a few people to help but putting it here too. https://gist.githubusercontent.com/sarina/9222b41ce12663723066/raw/5353cf007073a86a3540aaf355b22b1af35eac3c/gistfile1.txt

clrux replied:
It looks like both errors are caused from a missing font file - the
'edx-cc.svg' file. Let's make sure it's present. I believe I've still got
the .zip that contains all of the font files.

On Thu, Jan 8, 2015 at 9:25 AM, Sarina Canelake <notifications@github.com>
wrote:

> In common/static/css/edx-cc.css
> <https://github.com/edx/edx-platform/pull/6228#discussion-diff-22653637>:
>
> > @@ -0,0 +1,63 @@
> > +@font-face {
> > +  font-family: 'edx-cc';
> > +  src: url('../fonts/edx-cc/edx-cc.eot?52318265');
> > +  src: url('../fonts/edx-cc/edx-cc.eot?52318265#iefix') format('embedded-opentype'),
> > +       url('../fonts/edx-cc/edx-cc.woff?52318265') format('woff'),
> > +       url('../fonts/edx-cc/edx-cc.ttf?52318265') format('truetype'),
> > +       url('../fonts/edx-cc/edx-cc.svg?52318265#edx-cc') format('svg');
> > +  font-weight: normal;
> > +  font-style: normal;
> > +}
> > +/* Chrome hack: SVG is rendered more smooth in Windozze. 100% magic, uncomment if you need it. */
> > +/* Note, that will break hinting! In other OS-es font will be not as sharp as it could be */
>
> The sandbox keeps failing to build with the following 2 errors - I don't
> know how to resolve it. I've reached out to a few people to help but
> putting it here too.
> https://gist.githubusercontent.com/sarina/9222b41ce12663723066/raw/5353cf007073a86a3540aaf355b22b1af35eac3c/gistfile1.txt
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6228/files#r22653637>.
>



-- 
Chris Rodriguez | Front End Developer
141 Portland Street, 9th Floor
Cambridge, MA 02139
607-333-1667
crodriguez@edx.org
@clrux

sarina replied:
@clrux yes it is present in this PR

esbeka replied:
@sarina I modified the paths of the original edx-cc font when I added it to this PR. However, I must have missed the reference to the .svg version, it shows as a comment in my editor. I updated the reference in commit: 6175e3134066a9f1e0939f9d31a3d8c735192255.

sarina replied:
@esbeka cool thanks for the update. I'm trying a new sandbox deploy now.

explorerleslie commented:
Hi @esbeka, I'm reviewing the sandbox that @sarina set up. A few comments below, in order of priority: 

- When turning on the advanced setting on a course that previously had the Licensable setting set to FALSE, the default "All Rights Reserved" message doesn't display on the course or videos, even though when you go to the Schedule and Details page, the license is set to All Rights Reserved. It seems like I have switch to Creative Commons and then back to All Rights Reserved for the license to show up in Studio and LMS. 

- If I set the licenses for the course and some videos, and then turn the advanced setting OFF, the course licenses are hidden from Studio and the LMS, but the video licenses are still shown in LMS and Studio. 

- When setting a designation on the New Course, the checkbox icons aren't showing up, so I can't actually see the options I'm selecting: 
![image](https://cloud.githubusercontent.com/assets/5643426/5667291/991c73e8-9734-11e4-820d-f6cfd64a54fd.png)

- Similar issue on the video license section and the Schedules and Details page, but there's also some weird font/alignment issues too: 
![image](https://cloud.githubusercontent.com/assets/5643426/5667454/bc9c1ca0-9735-11e4-8db8-79d62d9371ae.png)

sarina commented:
@explorerleslie for the screenshots you took, can you please mention what browser and operating system you are using, in case it is a browser dependency issue?

explorerleslie commented:
@sarina yup. @esbeka, I'm using Chrome 39.0.2171.95 (64-bit) and Mac OS X Mavericks 10.10.1.

sarina commented:
@esbeka manually rerunning tests on our new build server: https://build.testeng.edx.org/job/edx-platform-all-tests/62/ (in case it doesn't report back to Github). 

It looks like you have a merge conflict. It's probably with AUTHORS. If you want to reduce the need for rebasing for conflicts, feel free to pull your AUTHORS changes into a separate pull request that we can just quickly merge.

sarina commented:
@esbeka you have a bunch of bok-choy test failures. Check out the console log for help tracking them down: https://build.testeng.edx.org/job/edx-platform-all-tests/62/SHARD=1,TEST_SUITE=bok-choy/artifact/test_root/log/bok_choy_studio.log

sarina commented:
@esbeka in addition to fixing tests, you'll definitely need to rebase. There's been some changes you need to pick up.

sarina commented on common/test/data/uploads/asset.html:
```diff
@@ -1,10 +1 @@
-</body>
-</html>
-
+test file
```
Why did you make this change?

sarina commented:
@esbeka I would like to understand how the license is specified in XML ([OLX authoring format](http://engineering.edx.org/2014/10/open-learning-xml-olx-format/)). We will need to document that, in addition to the user-facing changes, so people who do not use Studio will understand how to appropriately license their components.

marcotuts commented on lms/static/sass/shared/_course_object.scss:
```diff
@@ -267,3 +267,23 @@
 }
+
+.xmodule_licensable {
+  @include transition(opacity .25s ease-in-out);
```
NIT - I would use one of our tmg sass variables here instead of .25s specifically.

marcotuts commented on lms/static/sass/shared/_course_object.scss:
```diff
@@ -267,3 +267,23 @@
+  @include transition(opacity .25s ease-in-out);
+  clear: both;
+  padding: ($baseline / 4) 0;
+  opacity: .4;
```
NIT - s/opacity: .4;/opacity: 0.4;

sarina commented:
https://github.com/feedbackfruits/edx-platform/pull/2 is the result of doc review. Please merge when you can, in addition to addressing outstanding comments. Thanks

sarina commented:
@esbeka - will need another rebase - it had another conflict in AUTHORS, so I went ahead and added you and Steffan to AUTHORS in a separate PR to hopefully prevent too many conflicts going forward. Also there are a lot of comments in the PR that also need a response. I am looking at the failing tests now, sorry I did not get to this yesterday.

sarina commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -18,6 +18,7 @@
 from django.http import HttpResponseBadRequest, HttpResponse, Http404
 from django.utils.translation import ugettext as _
 from django.views.decorators.http import require_http_methods
+from django.conf import settings
```
Remove unused import

esbeka commented:
@sarina We squashed earlier commits and applied fixes based on your review comments and PR feedbackfruits/edx-platform#2. With these commits the license icons are displayed correctly, and the course license is shown underneath the course content inside the lms and studio.

To do list:
- [X] Course license doesn't show up after our latest rebase
- [X] Merge doc PR feedbackfruits/edx-platform#2
- [ ] Video license is still shown even though the course licensable is set to false
- [ ] Course rerun doesn't keep the license
- [ ] License selector is shown in the video editor even though the course licensable property is set to false

sarina commented:
Regarding the [last failing build](https://jenkins.testeng.edx.org/job/edx-all-tests-auto-pr/22071/) - the cms acceptance shard didn't start properly, and the bok choy test is flaky. I'm running a [new build here](https://build.testeng.edx.org/job/edx-platform-all-tests/177/)

sarina commented:
@singingwolfboy - today (Tuesday) could you please take a pass at reviewing this pull request as we discussed at standup this morning

singingwolfboy commented on cms/djangoapps/models/settings/course_metadata.py:
```diff
@@ -83,6 +84,11 @@ def fetch_all(cls, descriptor):
                 continue
+
+            # The licensable field should only be in the course settings if licensing is enabled globally.
+            if field.name is 'licensable' and not settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False):
```
Use `==` rather than `is` when comparing strings for equality.

singingwolfboy commented on cms/static/js/index.js:
```diff
@@ -51,12 +51,19 @@ define(["domReady", "jquery", "underscore", "js/utils/cancel_on_escape", "js/vie
             var number = $newCourseForm.find('.new-course-number').val();
             var run = $newCourseForm.find('.new-course-run').val();
+            var license = {};
+            if(licenseSelector){
```
Are you checking to see if the `licenseSelector` variable is defined? Because if so, this will fail with a `ReferenceError` if the variable is not defined. (If the variable is defined, but set to a falsy value like `null`, then it will do what you expect.) If you need to worry about the case where the variable may not be defined, then you're better off using the `typeof` keyword to check that.
```js
if (typeof licenseSelector != "undefined" && licenseSelector) {
```

singingwolfboy commented on cms/static/js/models/license.js:
```diff
@@ -0,0 +1,51 @@
+     * CC-BY-*  - Creative Commons Attribution license
+     * CC-*-ND  - Creative Commons NoDerivatives license
+     * CC-*-NC  - Creative Commons NonCommercial license
+     * CC-*-SA  - Creative Commons ShareAlike license
```
Is this model written to allow for new, unexpected license types? Or does this model expect that these are the *only* possible licenses that can exist? You should document this one way or the other. If you are writing your model to allow for new, unexpected license types, be sure to write a test case that uses a fake license type and assert that things still work.

singingwolfboy commented on cms/static/js/models/license.js:
```diff
@@ -0,0 +1,51 @@
+            else {
+                newattrs.kind = "NONE";
+            }
+            if (!_.isEmpty(errors)) return errors;
```
Please use curly braces for readability and consistency. You don't strictly need them here, but it's confusing to have an `if` clause without curly braces, and this code will probably be edited in the future to make the necessary. Simpler to just include them now.

singingwolfboy commented on cms/static/js/models/license.js:
```diff
@@ -0,0 +1,51 @@
+                newattrs.kind = validKind;
+            }
+            else {
+                newattrs.kind = "NONE";
```
Why not set `newattrs.kind` to `null` or `false`, instead? `"NONE"` is a truthy value, which may cause problems.

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+                }
+                else if (!(this.model instanceof LicenseModel) && this.model.kind) {
+                    this.model = new LicenseModel(this.model);
+                }
```
Whoa whoa whoa. This is going to be *incredibly* confusing for anyone trying to debug this code. You pass in a model, and the view uses a different model than the one you passed in? That's going to cause all sorts of unexpected situations and behaviors.

Instead, I recommend one of two approaches. I think the best option is to just ignore this type check, and assume that the model has the attributes that this view is expecting. This is called [duck typing](https://en.wikipedia.org/wiki/Duck_typing), and it works surprisingly well. It also means that people can write new models that match the `LicenseModel` API, plug them into this view, and they will Just Work. If you don't want to use duck typing, then I'd suggest that you simply throw an error if the model that is passed in is not the model that you're expecting. Let the developer deal with that by wrapping the model herself before passing it to the view, so that she knows that the wrapping is happening.

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+            initialize: function(options) {
+                this.template = this.loadTemplate("license-selector");
+                if (!this.model) {
+                    this.model = new LicenseModel();
```
I don't think it's worth it to have a default model like this. If you instantiate this view without passing a model to it, something is *wrong*, and that should be detected as early as possible. If you really want to support that case, then you better have several unit tests that cover that case.

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+                    license_preview: this.licenseView.renderLicense()
+                }));
+
+                this.$el.addClass('license-selector');
```
Shouldn't this go into the template, rather than being added here?

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+
+                if (!kind || kind === "NONE" || kind === "ARR") {
+                    this.$el.find('.license-button[data-license="ARR"]').addClass('selected');
+                    this.$el.find('.license-button[data-license="CC"]').removeClass('selected');
```
These should go into the template.

esbeka replied:
It shouldn't, as the selected state should only be added when the license kind is ARR. This method is called every time the license changes, to update the view and selected elements.

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+                    var attr = kind.split("-");
+                    this.$el.find('.license-button').removeClass('selected');
+                    for(i in attr) {
+                        this.$el.find('.license-button[data-license="' + attr[i] + '"]').addClass('selected');
```
This should go into the template.

esbeka replied:
same here, this is necessary to show the license attributes in selected state when the license is set.

singingwolfboy commented on cms/static/js/views/license-selector.js:
```diff
@@ -0,0 +1,107 @@
+                    $button.toggleClass("selected");
+
+                    if (buttonLicenseKind === "ND" && $button.hasClass("selected")) {
+                        $cc.children(".license-button[data-license='SA']").removeClass("selected");
```
Do not modify the HTML like this. You want to set and modify attributes on the model, which will trigger this view to re-render itself. That's the whole point of the MVC pattern.

esbeka replied:
I agree, will update this.

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                if (!this.model) {
+                    this.model = new LicenseModel();
+                }
+                else if (!(this.model instanceof LicenseModel)) {
```
See earlier comment about the view altering the model that was passed in. Very, very bad idea. Don't do this.

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+            },
+
+            render: function() {
+                this.$el.html(this.renderLicense());
```
Why not just put the contents of the `renderLicense` function into this `render` function, and delete the `renderLicense` function?

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                else if(kind === "CC0"){
+                    // Creative commons zero license
+                    licenseText = gettext("No rights reserved")
+                    return "<a rel='license' href='http://creativecommons.org/publicdomain/zero/1.0/' target='_blank'><i class='icon-cc'></i><i class='icon-cc-zero'></i><span class='license-text'>" + licenseText + "</span></a>";
```
This might be better as a template.

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                else {
+                    // Creative commons license
+                    licenseVersion = "4.0";
+                    licenseHtml = "<i class='icon-cc'></i>";
```
This would *definitely* be better as a template.

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                    licenseHtml = "<i class='icon-cc'></i>";
+                    licenseLink = [];
+                    licenseText = [];
+                    if(/BY/.exec(kind)){
```
For `if` tests, you want to use [`RegExp.test`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test), not [`RegExp.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec).

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                        // For example, see: http://creativecommons.org/licenses/by-nc-nd/4.0/
+                        licenseText.push(gettext("NonDerivatives"));
+                    }
+                    licenseTooltip = interpolate(gettext("This work is licensed under a Creative Commons %(license_attributes)s %(version)s International License."), {
```
This will definitely need a translators comment.

singingwolfboy commented on cms/static/js/views/license-view.js:
```diff
@@ -0,0 +1,106 @@
+                            license_attributes: licenseText.join("-"),
+                            version: licenseVersion
+                        }, true);
+                    return "<a rel='license' href='http://creativecommons.org/licenses/" +
```
This needs to be a template so badly that it's screaming out in pain.

singingwolfboy commented on cms/static/js/views/metadata.js:
```diff
@@ -118,6 +120,43 @@ function(BaseView, _, MetadataModel, AbstractEditor, FileUpload, UploadDialog, V
+                    model: new LicenseModel(this.model.getValue()),
+                    el: this.$el.find('.wrapper-license-selector')
+                });
+                this.initialized = true;
```
Shouldn't this be set in the `initialize` method?

singingwolfboy commented on cms/static/js/views/settings/main.js:
```diff
@@ -26,6 +26,12 @@ var DetailsView = ValidatingView.extend({
+        if(!this.model.has('license')){
+            this.model.set('license', {kind: 'ARR', version: '1.0'});
+        }
+        this.linkLicense();
```
Why don't you have the `linkLicense` method listen to the model, and call itself automatically when the `license` attribute changes?

singingwolfboy commented on cms/templates/course_outline.html:
```diff
@@ -132,6 +132,13 @@ <h3 class="title-3">${_("Changing the content students see")}</h3>
+        % if settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False) and context_course.licensable and context_course.license:
+          <div class="course-license">
+              <div class="xmodule_licensable license">
+                ${context_course.license.html}
```
You should not have a method generating HTML that you can dump into the template like this. The HTML should be generated in the template. That's what templates are for.

singingwolfboy commented on cms/templates/settings_advanced.html:
```diff
@@ -5,6 +5,7 @@
   from django.utils.translation import ugettext as _
   from contentstore import utils
   from django.utils.html import escapejs
+  from django.conf import settings
```
What is this for?

singingwolfboy commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -634,6 +636,20 @@ class CourseFields(object):
+    licensable = Boolean(
+        display_name=_("Licensable"),
+        help=_("Whether this course and its contents can be licensed using Creative Commons Licensing."),
+        default=hasattr(settings, 'FEATURES') and settings.FEATURES.get("DEFAULT_COURSE_LICENSABLE", False),
```
I think we can safely assume that there will always be a `FEATURES` dict in the settings object. We assume that in many other places.

singingwolfboy commented on common/lib/xmodule/xmodule/license.py:
```diff
@@ -0,0 +1,168 @@
+
+        This method should be overridden in child classes to provide the desired html.
+        """
+        return u"<p>" + _("This resource is not licensed.") + u"</p>"
```
Don't do this. HTML should be generated in a template, not in a method or property.

singingwolfboy commented on common/lib/xmodule/xmodule/license.py:
```diff
@@ -0,0 +1,168 @@
+        """
+        return u"<p>" + _("This resource is not licensed.") + u"</p>"
+
+    def to_json(self, value):
```
This is a method on a class. It should return the JSON representation of the instance it is called on. It should not return the JSON representation of the `value` argument. It shouldn't even accept a `value` argument.

singingwolfboy commented on common/lib/xmodule/xmodule/license.py:
```diff
@@ -0,0 +1,168 @@
+        super(ARRLicense, self).__init__(kind, version, *args, **kwargs)
+
+    @property
+    def html(self):
```
Use a template to generate HTML, not a method.

singingwolfboy commented on common/lib/xmodule/xmodule/license.py:
```diff
@@ -0,0 +1,168 @@
+            self.version = "4.0"
+
+    @property
+    def html(self):
```
Use a template to generate HTML, not a method.

singingwolfboy commented on common/lib/xmodule/xmodule/license.py:
```diff
@@ -0,0 +1,168 @@
+            version=self.version
+        )
+
+def parse_license(kind_or_license, version=None):
```
This whole method seems like a bad idea. Why do we want to allow licenses to exist in any of several different representations or states? Seems to me like we should expect a license to be represented in exactly one way.

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -1189,5 +1190,7 @@ def default(self, obj):
             else:
                 return obj.isoformat()
+        elif isinstance(obj, License):
+            return obj.to_json(obj)
```
Why wouldn't you just call `obj.to_json()`?

singingwolfboy commented on common/lib/xmodule/xmodule/tests/test_license.py:
```diff
@@ -0,0 +1,136 @@
+        self.invalidDict = {"kind": "RAND"}
+
+    def test_html(self):
+        """Should never be called in production, but otherwise say it's not licensed."""
```
Why should `License().html` never be called in production?

htdvisser replied:
No idea why I added that comment. > Removed.

singingwolfboy commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -375,6 +392,10 @@ def editable_metadata_fields(self):
         else:
             editable_fields.pop('source')
 
+        if not(hasattr(settings, 'FEATURES') and settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False)):
```
I think we can assume that there is a `settings.FEATURES` dict.

singingwolfboy commented on lms/djangoapps/courseware/courses.py:
```diff
@@ -234,6 +235,23 @@ def get_course_about_section(course, section_key):
 
+def get_course_license_section(course):
+    """
+    This returns the snippet of html to be rendered on the course about page under the license section,
```
HTML should be rendered in templates, not in methods.

singingwolfboy commented on lms/templates/courseware/course_about.html:
```diff
@@ -218,6 +218,9 @@
       <div class="inner-wrapper">
         ${get_course_about_section(course, "overview")}
+        % if settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False) and course.licensable:
+          ${get_course_license_section(course)}
```
HTML should be written into this template, not generated in the `get_course_license_section()` function.

singingwolfboy commented on lms/templates/courseware/course_footer.html:
```diff
@@ -0,0 +1,7 @@
+<div class="container course-footer">
+  % if settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False) and course.licensable and course.license:
+    <div class="xmodule_licensable">
+      ${course.license.html}
```
HTML should be written into this template, not generated in the `license.html` attribute.

singingwolfboy commented on lms/templates/video.html:
```diff
@@ -141,5 +145,10 @@ <h3 class="hidden">${_('No playable video sources found.')}</h3>
     % endif
+    % if settings.FEATURES.get('CREATIVE_COMMONS_LICENSING', False) and licensable and video_license:
+      <div class="xmodule_licensable license">
+        ${video_license.html}
```
HTML should be written into this template, not generated in the `video_license.html` attribute.

singingwolfboy commented:
This pull request doesn't have any Jasmine tests at all. That's a problem.

sarina commented:
We are going to take this work over internally, so closing this pull request. It is forked over at #6823 with some additional work to try to use inheritable fields, which we will continue jamming on.
